includes/Jobs/GenerateProductFeed.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace SkyVerge\WooCommerce\Facebook\Jobs;
+
+use Automattic\WooCommerce\ActionSchedulerJobFramework\AbstractChainedJob;
+use Automattic\WooCommerce\ActionSchedulerJobFramework\Utilities\BatchQueryOffset;
+use Exception;
+use WC_Facebookcommerce;
+use WP_Query;
+
+/**
+ * Class GenerateProductFeed
+ *
+ * @since 2.5.0
+ */
+class GenerateProductFeed extends AbstractChainedJob {
+
+	use BatchQueryOffset;
+
+	/**
+	 * Called before starting the job.
+	 */
+	protected function handle_start() {
+		// Optionally override this method in child class.
+	}
+
+	/**
+	 * Called after the finishing the job.
+	 */
+	protected function handle_end() {
+		// Optionally override this method in child class.
+	}
+
+	/**
+	 * Get a set of items for the batch.
+	 *
+	 * NOTE: when using an OFFSET based query to retrieve items it's recommended to order by the item ID while
+	 * ASCENDING. This is so that any newly added items will not disrupt the query offset.
+	 *
+	 * @param int   $batch_number The batch number increments for each new batch in the job cycle.
+	 * @param array $args         The args for the job.
+	 *
+	 * @throws Exception On error. The failure will be logged by Action Scheduler and the job chain will stop.
+	 */
+	protected function get_items_for_batch( int $batch_number, array $args ): array {
+		$product_args = [
+			'fields'         => 'ids',
+			'post_status'    => 'publish',
+			'post_type'      => [ 'product', 'product_variation' ],
+			'posts_per_page' => $this->get_batch_size(),
+			'offset'         => $this->get_query_offset( $batch_number ),
+			'orderby'        => 'ID',
+			'order'          => 'ASC',
+		];
+
+		$query = new WP_Query( $product_args );
```

Comment:
Fetching by id felt much faster - but this is not something I have measured.

Reply:
I would argue that the reading option with prepared items not only solves the published variation parents issue but it would be faster than doing a `new WP_Query` every time.
What we could do to optimize a bit is to have the batch number in a separate option so we would only need to read the option once and we would store the batch number after batch processing.

Reply:
This is something we'll have to test I think. I just know when we used this approach of doing large option table read/writes for each batch when using AW we had user complaints.

> What we could do to optimize a bit is to have the batch number in a separate option so we would only need to read the option once and we would store the batch number after batch processing.

Do you mean we would only need to "write" the option once?

Reply:
Reading big options is costly and writing it is costly. If we would go with ids in an option approach we could use a separate option for tracking the current batch number. This way, we only need to read the big option and we don't need to write it just to update the batch number.

Reply:
> If we would go with ids in an option approach we could use a separate option for tracking the current batch number.

Yeah avoiding writes, might be faster.

> This way, we only need to read the big option and we don't need to write it just to update the batch number.

Actually we don't need to store the batch number as an option since it's passed as an argument to the scheduled action by the job framework.

Reply:
Even better :)